# Python  Boot Camp 2
This is a continuation of [Python Boot Camp 1](python_boot_camp_1.ipynb). Based on [MATLAB Onramp](https://matlabacademy.mathworks.com)

### Required  packages
Run the %pylab inline magic below to get started

In [2]:
%pylab inline

Populating the interactive namespace from numpy and matplotlib


## 3. 1D and 2D Arrays
Create Python variables that contain multiple elements.

### Python Data Types

Python has a number of [built-in data types](https://www.w3schools.com/python/python_datatypes.asp), in the following categories:
 
 - Text Type:	string, abbreviated `str`
 - Numeric Types:	integer `int`, floating point `float`, `complex`
 - Sequence Types:	`list`, `tuple`, `range`
 - Mapping Type:	dictionary `dict`
 
Of these, we will mostly use the `str`, `int`, `float`, and `list` types.  The other types can also be useful though:

 - The `tuple` type is like a list, except it is immutable, meaning it cannot be changed once defined.
 - The `dict` type allows for storage of data consisting of key value pairs, making it a mini database.

### Python lists

In Python, a `list` is a collection of multiple elements, separated by commas and enclosed in square brackets. **Note that, unlike MATLAB, the comma separator is _required_.**

```
In  [1]: p = [3, 'film']
``` 
> **TASK:** Create a list named x with two elements: 7 and 9

<hr>

The problem with lists is that we can't do math on them.

> **TASK:** Show that adding 5 to x results in an error.

<hr>

`numpy` contains a function `array` that converts lists of numbers to a form that we can use to do math.

> **TASK:** Create an array `z` from `x` using `array(x)`

<hr>

> **TASK:** add 5 to every element of array z

<hr>

> **TASK:** Create an array named x that contains the values 3, 10, and 5 in that order.

<hr>

We can make a two-dimensional array using a list of lists:

```
In  [1]: x = array( [ [3, 4, 5], [6, 7, 8] ] )
         x
Out [1]: array([[3, 4, 5],
                [6, 7, 8]])
```
> **TASK:** Create a 2D array named x with the values shown below.
```
5    6    7
8    9   10
```

<hr>

In Python, we can perform calculations within the square brackets.

```
In  [1]: x = array([abs(-4), 4**2])
         x
Out [1]: array([ 4, 16])
```

**Note:** Python uses a double-asterisk `**` as the symbol for exponents, not the carat `^` as in MATLAB. 

> **TASK:** Create a 1D array named `x` that contains `sqrt(10)` as its first element and π$^2$ as its second element.

### Creating Evenly-Spaced Arrays

It is common to create arrays containing evenly-spaced numbers. For large arrays, entering individual numbers is not practical. An alternative, shorthand way to create evenly-spaced arrays is to use numpy's `arange` function:
```
In  [1]: x = arange(4, 12, 2)
         x
Out [1]: array([ 4,  6,  8, 10])
```

Note that `4` is the starting value, `12` is the first _excluded_ value, and `2` is the step size. This function is the array version of the function `range` that is built into Python.

> **TASK:** Create an array named x with values 1, 2, 3, and 4, using the `arange` function.

### Creating arrays with `linspace`

If the step size is a non integer, such as 0.1, it is best to use `numpy.linspace`. 

```
In  [1]: y = linspace(0.4,3,6)
         y
Out [1]: array([0.4 , 0.92, 1.44, 1.96, 2.48, 3.  ])
```

Here, `0.4` is the starting value, `3` is the last _included_ value, and `6` is the number of elements in the array. This function is equivalent to the `linspace` function in MATLAB.

> **TASK:** Create an array named x that starts at 1, ends at 10, and contains 5 elements.

### Array Creation Functions

`numpy` includes functions to create commonly used matrices, such as matrices of random numbers.
```
In  [1]: x = random_sample([2,2])
         x
Out [2]: array([[0.78363163, 0.07238245],
                [0.68659162, 0.17798821]])
```
Note that the inputs to the function `random_sample` is a list representing the dimensions of the random array.
> **TASK:** Create a variable named x that is a 5-by-4 array of random numbers.

<hr>

> **TASK:** Use the `zeros` function to create an array of all zeros that has 6 rows and 3 columns (6-by-3). Assign the result to a variable named x.

<hr>

#### Further Practice:
 - What do you think the function `numpy.ones` does?
 - How do you get the size of an existing array? You can use `numpy.shape`.
 - You can also create an array with the same size as an existing array using `numpy.zeros_like` or `numpy.ones_like`.

## 4. Loading and Saving Data
We don't often save data to files in Python, but we may import data from an external source. Let's load some data from a "comma separated value" or CSV file using `np.loadtxt`. As inputs, this function requires the file name or path as a text string, and we need to specify the `delimiter` between values to be a comma.   
```
In  [1]: data = loadtxt('data.csv', delimiter=',')
``` 
> **TASK:** Load the data in `data.csv` into an array named `data` and show that the `shape` of this array is `(40, 19)`

In [4]:
data = loadtxt('data.csv', delimiter=',')
data

array([[ 28.        ,  34.5       ,  15.        ,  29.        ,
         15.5       ,  20.        ,  16.66666667,  33.        ,
          8.62083333,  10.        ,  14.        ,  18.5       ,
         18.        ,  18.        ,   9.08333333,  10.        ,
         99.        ,  73.        ,  21.5       ],
       [ 27.5       ,  33.5       ,  14.        ,  29.        ,
         19.        ,  28.        ,  16.66666667,  31.        ,
          9.0875    ,  10.        ,  21.        ,  19.        ,
         21.        ,  21.        ,  10.5       ,  10.        ,
        106.        ,  87.        ,  24.125     ],
       [ 27.        ,  27.        ,  11.        ,  27.        ,
         18.5       ,  26.5       ,  15.33333333,  27.5       ,
          8.6125    ,  10.        ,  19.        ,  15.        ,
         21.        ,  21.        ,  10.16666667,  10.        ,
         97.        ,  90.        ,  23.375     ],
       [ 27.        ,  25.        ,  16.        ,  26.5       ,
         13.   

## 5. Indexing into and Modifying Arrays
Use indexing to extract and modify rows, columns, and elements of `numpy` arrays.

### Indexing into Arrays

You can extract values from an array using row, column indexing.

```
In  [1]: x =  A[5, 7]
``` 

This syntax extracts the value in the 5th row and 7th column of `A` and assigns the result to the variable `y`. A couple of important notes:

 - Indexing into arrays and lists is done using square brackets with comma separators, `[5, 7]`.
 - For a 2D array, the first index is always the row number and the second index the column number, just like MATLAB.
 - **Python indexing starts with 0 (zero) and so the beginning row is the `0th` row and the `5th` row has 5 rows above it!!**
 
> **TASK:** Create a variable `x` that contains the value in the `0th` row and `3rd` column of the array `data`.

<hr>

You can use the index `-1` as either a row or column index to reference the last element.
```
In  [1]: y = A[-1,2]
``` 

> **TASK:** Use the `-1` index to obtain the value in the last row and 4th column of the array `data`. Assign this value to a variable named x.

<hr>

We can access elements close to the end of an array using larger negative indices.
```
In  [1]: y =  A[-3,-2]
``` 

> **TASK:** Create a scalar variable `x` that contains the value in the second to last row and 3rd to last column of `data`.

<hr>

#### Further Practice
If you only use one index with a 2D array, it will return the row associated with that index. Using one index, try extracting the eighth row of `data`.

You can also use variables as your index. Try creating a variable `y`, and use `y` as the index to `data`.

### Extracting Multiple Elements

When used as an index, the colon operator (:) specifies all the elements in that dimension. The syntax
 ```
 x = A[2,:]
 ```
creates a 1D array containing all of the elements from the `2nd` row of A. In this case both `A[2]` and `A[2,:]` produce the same result.

> **TASK:** Create a variable named `density` that contains the second column of the 2D array named `data`.

<hr>

The colon operator can refer to a range of values. The following syntax creates a 2D array containing the `0th` through `2nd` columns of the array `A`.

```
y =  A[:, 0:3]
``` 

**Note:** The value `3` represents the first _excluded_ column extracted from `A`. If there is no excluded column (*i.e.* we want to include column `-1`) we can omit an index, for example:
```
y =  A[:, 0:]
``` 

> **TASK:** Create a variable `volume` containing the last two columns of `data`.

<hr>

`1D` arrays require a single index. For example: 

```
u =  v[4]
``` 

returns the `4th` element of 1D array `v`.
> **TASK:** Using a single index value, create a variable named `p` containing the `6th` element in the 1D array `density`.

<hr>

A single range of index values can be used to reference a subset of a 1D array. For example: 

```
x = v[3:]
```

returns a subset of 1D array `v` from the `3rd` element to the end.

> **TASK:** Using a range of index values, create a 1D array named `p` containing the `2nd` through `5th` elements of `density`.

<hr>

#### Further Practice
Indices can be lists of non-consecutive numbers, such as `[4, 8, 3]` Try extracting the first, third, and sixth elements of density.

In [10]:
density = array([1, 4, 6, 3, 5, 8, 9, 10, 5, 3, 1])
density[[0, 3, 4]]

array([1, 3, 5])

### Changing Values in Arrays

Task 1
Remember you can use the : character to extract entire columns of data.
> **TASK:** Create a 1D array named `v1` containing the last column of `data`.

<hr>

It's very important to note that `v1` is not independent of `data`. If we change `v1`, we change `data`. We can alter elements of an array by combining indexing with assignment:
```
A[2] = 11
```

> **TASK:** Change the `0th` element in `v1` from 21.5 to 0.5. Then, show that `data[0,-1]` also changes.

<hr>

> **TASK:** Change the value of the element in the first row and last column of data back to 21.5. What happens to `v1[0]`?

<hr>

If we want to make `v1` a separate, independent copy, we can use the `numpy.copy` function.

```
y= copy(A[:,0])
```

> **TASK:** Create a 1D array named `v2` that contains the last column of `data`, by using the `copy` function. Then, change the value of the `0th` element of `v2` to `0.5` and show that `data[0,-1]` remains 21.5.

<hr>

#### Further Practice

You can combine indexing with assignment to change array values to equal other elements. For example, this code would change the value of x[1] to x[2]:
```
x[1] = x[2]
```

Try changing the `0th` column of `data` to the `1st` column of `data`.

### Unpacking Lists and Arrays

Both lists and arrays can be unpacked into individual variables:

```
k1, k2, k3, k4 = [0.1, 0.2, 0.3, 0.5]
```

> **TASK:** Create a 1D random array of length 3 using `random_sample` and unpack it into three variables, `r1`, `r2`, `r3`

In [22]:
dogs = ['rover', 'bowser', 'ruff']
dogs

['rover', 'bowser', 'ruff']

In [23]:
cats = ['fifi', 'heidi', 'riley']
cats

['fifi', 'heidi', 'riley']

In [27]:
pets =[dogs, cats]
pets
array(pets)

array([['rover', 'bowser', 'ruff'],
       ['fifi', 'heidi', 'riley']], dtype='<U6')

In [33]:
row1 = [ 1, 2, 3]
row2 = [3, 4, 5]
a = [row1, row2]
b= array(a)
b

array([[1, 2, 3],
       [3, 4, 5]])

In [37]:
b[:, -1]

array([3, 5])

In [31]:
numpy.array(row1) + 10

array([11, 12, 13])

In [21]:
a, random_sample([2,3])

(array([0.95111814, 0.94662348, 0.63184258]),
 array([[0.24992489, 0.3688355 , 0.75421645],
        [0.41724138, 0.75573344, 0.64651401]]))

## 6. Array Calculations
Perform calculations on entire arrays at once.

### Performing Array Operations on 1D Arrays

`numpy` is designed for natural math on arrays. For example, you can add a scalar value to all the elements of an array.

```
y = x + 2
```

> **TASK:** Add 1 to each element of `v2` and store the result in a variable named `r`.

<hr>

You can add together any two arrays of the same size.

```
z = x + y
```

> **TASK:** Create a 1D array `vavg` that is the average of the arrays `v1` and `v2`.

<hr>

Basic statistical functions in Python can be applied to an array to produce a single output. The maximum value of a 1D array can be determined using the `np.max` function.
```
xMax = max(x)
```
> **TASK:** Create a variable vm containing the maximum of the array `vavg`.

<hr>

In `numpy`, the `*` operator performs element-wise array multiplication, as opposed to matrix multiplication. You can use * to multiply two equally sized arrays. 

```
In  [1]: z = array([3, 4]) * array([10, 20])
         z 
Out [1]: 30    80
```
**NOTE:** Using `numpy`, the "dot" operators `.*` and `./` are not required to accomplish element-wise math, as they are in MATLAB.

> **TASK:** Create a variable named `mass` containing the elementwise product of `density` and `vavg`.

<hr>

#### Further Practice
You have performed array operations with arrays of the same size and with scalars and arrays.
The concept of [broadcasting](https://numpy.org/doc/stable/user/basics.broadcasting.html) enables other compatible sizes. For example, try the following:

```
x = array([[1, 3, 5, 7], [2, 4, 6, 8]]) * array([1,2,3,4])
```

What size is x?

<hr style="height:4px;border-width:0;color:gray;background-color:gray">

This is the end of **Python Boot Camp 2**. Be sure to save by pressing the "S" key in command mode, and continue to [**Python Boot Camp 3**](python_boot_camp_3.ipynb)